src/hashtable.h

```diff
@@ -0,0 +1,167 @@
+     * comparing the pointers for equality. */
+    int (*keyCompare)(const void *key1, const void *key2);
+    /* Callback to free an entry when it's overwritten or deleted.
+     * Optional. */
+    void (*entryDestructor)(void *entry);
+    /* Callback to control when resizing should be allowed. */
+    int (*resizeAllowed)(size_t moreMem, double usedRatio);
+    /* Invoked at the start of rehashing. */
+    void (*rehashingStarted)(hashtable *ht);
+    /* Invoked at the end of rehashing. */
+    void (*rehashingCompleted)(hashtable *ht);
+    /* Track memory usage using this callback. It is called with a positive
+     * number when the hashtable allocates some memory and with a negative number
+     * when freeing. */
+    void (*trackMemUsage)(hashtable *ht, ssize_t delta);
+    /* Allow a hashtable to carry extra caller-defined metadata. The extra memory
+     * is initialized to 0. */
+    size_t (*getMetadataSize)(void);
+    /* Flag to disable incremental rehashing */
+    unsigned instant_rehashing : 1;
+} hashtableType;
+
+typedef enum {
+    HASHTABLE_RESIZE_ALLOW = 0,
+    HASHTABLE_RESIZE_AVOID,
+    HASHTABLE_RESIZE_FORBID,
+} hashtableResizePolicy;
+
+typedef void (*hashtableScanFunction)(void *privdata, void *entry);
+
+/* Constants */
+#define HASHTABLE_BUCKET_SIZE 64 /* bytes, the most common cache line size */
+
+/* Scan flags */
+#define HASHTABLE_SCAN_EMIT_REF (1 << 0)
+
+/* --- Prototypes --- */
+
+/* Hash function (global seed) */
+void hashtableSetHashFunctionSeed(const uint8_t *seed);
+uint8_t *hashtableGetHashFunctionSeed(void);
+uint64_t hashtableGenHashFunction(const char *buf, size_t len);
+uint64_t hashtableGenCaseHashFunction(const char *buf, size_t len);
+
+/* Global resize policy */
+void hashtableSetResizePolicy(hashtableResizePolicy policy);
+
+/* Hashtable instance */
+hashtable *hashtableCreate(hashtableType *type);
+void hashtableRelease(hashtable *ht);
+void hashtableEmpty(hashtable *ht, void(callback)(hashtable *));
+hashtableType *hashtableGetType(hashtable *ht);
+void *hashtableMetadata(hashtable *ht);
+size_t hashtableSize(hashtable *ht);
+size_t hashtableBuckets(hashtable *ht);
+size_t hashtableChainedBuckets(hashtable *ht, int table);
+size_t hashtableMemUsage(hashtable *ht);
+void hashtablePauseAutoShrink(hashtable *ht);
+void hashtableResumeAutoShrink(hashtable *ht);
+int hashtableIsRehashing(hashtable *ht);
+int hashtableIsRehashingPaused(hashtable *ht);
+void hashtableRehashingInfo(hashtable *ht, size_t *from_size, size_t *to_size);
+int hashtableRehashMicroseconds(hashtable *ht, uint64_t us);
+int hashtableExpand(hashtable *ht, size_t size);
+int hashtableTryExpand(hashtable *ht, size_t size);
+int hashtableExpandIfNeeded(hashtable *ht);
+int hashtableShrinkIfNeeded(hashtable *ht);
+hashtable *hashtableDefragTables(hashtable *ht, void *(*defragfn)(void *));
+
+/* Entries */
+int hashtableFind(hashtable *ht, const void *key, void **found);
```

Comment:
Interesting idea.

I would call it `hashtableEntryRef { void *entry; }` then and it doesn't need to be opaque.

The struct would need to be (stack) allocated by the caller, right? How would you change `void **hashtableFindRef(hashtable *ht, const void *key)`?

Rain suggested that instead of casting to `void **`, we can call these functions with a pointer to a `void *` and cast afterwards (which IIUC also has the benefit that we're not breaking the strict aliasing rules if we're storing non-pointers) like this:

```C
    void *entry;
    if (hashtableFind(ht, key, &entry)) {
        robj *obj = entry;
        ...
    }
```


Jim suggested that we use the opaque `hashtablePosition` for this, and provide a function to get the entry from it. The accessor function could also check that the position is still valid, using an epoch counter in the hashtable struct to detect changes. I started to refactor to that, but it became a massive refactoring and I didn't get it to work. So maybe later, as a follow up. But I can try your

Reply:
I've changed as the code example to avoid casting to `void **`.